distribution/lib/Standard/Base/0.0.0-dev/src/Data/Pair.enso

```diff
@@ -53,9 +66,189 @@ type Pair
+    index_of self element start=0 = check_start_valid start used_start->
+        predicate = case element of
+            _ : Function -> element
+            _ -> (==element)
+        if used_start<1 && predicate self.first then 0 else
+            if used_start<2 && predicate self.second then 1 else
+                Nothing
+
+    ## Returns the last index of an element in the pair.
+       Returns Nothing if the element is not found.
+
+       Arguments:
+       - element: The element to search for or a predicate function to test for
+         each element.
+       - start: The index to start searching backwards from. If the index is
+         negative, it is counted from the end of the pair.
+
+       > Example
+         Find the last index of an element in a pair.
+
+             Pair.new 2 2 . last_index_of 2 == 1
+    last_index_of : (Any | (Any -> Boolean)) -> Integer -> Integer | Nothing
+    last_index_of self element start=-1 = check_start_valid start max=2 used_start->
+        predicate = case element of
+            _ : Function -> element
+            _ -> (==element)
+        if used_start>0 && predicate self.second then 1 else
+            if used_start>=0 && predicate self.first then 0 else
+                Nothing
+
+    ## Combines all the elements of the pair, by iteratively applying the
+       passed function with next elements of the pair.
+
+       Arguments:
+       - init: The initial value for the fold.
+       - function: A function taking two elements and combining them.
+
+       > Example
+         Compute the sum of all of the elements in a pair.
+
+             Pair.new 10 20 . fold 0 (+)
+    fold : Any -> (Any -> Any -> Any) -> Any
+    fold self init function =
+        function (function init self.first) self.second
+
+    ## Combines all the elements of the pair using a binary operation.
+
+       Arguments:
+       - function: A binary operation that takes two items and combines them.
+
+       > Example
+         Compute the sum of all the elements in a Pair.
+
+             Pair.new 10 20 . reduce (+)
+    reduce : (Any -> Any -> Any) -> Any
+    reduce self function =
+        function self.first self.second
+
+    ## Reverses the pair, returning a pair with the same elements, but in
+       the opposite order.
+
+       > Example
+         Reverse a two-element vector.
+
+             Pair.new 1 2 . reverse
+    reverse : Pair
+    reverse self = Pair.new self.second self.first
+
+    ## Applies a function to each element of the pair.
+
+       Unlike `map`, this method does not return the individual results,
+       therefore it is only useful for side-effecting computations.
+
+       Arguments:
+       - function: A function to apply to each element of the pair.
+
+       > Example
+         Print each element in the vector to standard output.
+
+             Pair.new 1 2 . each IO.println
+    each : (Any -> Any) -> Nothing
+    each self f =
+        f self.first
+        f self.second
```

Comment:
```suggestion
    each self f =
        f self.first
        f self.second
        Nothing
```

In most of our cases `each` is returning `Nothing` and I'd say it's part of its 'contract', so let's be consistent there, instead of returning the second function's result.